server/src/main/java/org/elasticsearch/indices/recovery/plan/ShardSnapshotsService.java

```diff
@@ -118,4 +142,84 @@ public void fetchLatestSnapshotsForShard(ShardId shardId, ActionListener<Optiona
     protected boolean masterSupportsFetchingLatestSnapshots() {
         return clusterService.state().nodes().getMinNodeVersion().onOrAfter(SNAPSHOT_RECOVERIES_SUPPORTED_VERSION);
     }
+
+    private static final class InMemoryDirectory extends BaseDirectory {
+        private final Map<String, StoreFileMetadata> files;
+
+        private InMemoryDirectory(Map<String, StoreFileMetadata> files) {
+            super(new SingleInstanceLockFactory());
+            this.files = files;
+        }
+
+        @Override
+        public String[] listAll() {
+            return files.keySet().toArray(new String[0]);
+        }
+
+        @Override
+        public IndexInput openInput(String name, IOContext context) throws IOException {
+            StoreFileMetadata storeFileMetadata = getStoreFileMetadata(name);
+            if (storeFileMetadata.hashEqualsContents() == false) {
+                throw new IOException("Unable to open " + name);
+            }
+
+            final BytesRef data = storeFileMetadata.hash();
+            return new ByteArrayIndexInput(name, data.bytes, data.offset, data.length);
+        }
+
+        @Override
+        public void deleteFile(String name) {
+            throw new UnsupportedOperationException("this directory is read-only");
+        }
+
+        @Override
+        public long fileLength(String name) throws IOException {
+            final StoreFileMetadata storeFileMetadata = getStoreFileMetadata(name);
+            return storeFileMetadata.length();
+        }
+
+
+        @Override
+        public IndexOutput createOutput(String name, IOContext context) {
+            throw new UnsupportedOperationException("this directory is read-only");
+        }
+
+        @Override
+        public IndexOutput createTempOutput(String prefix, String suffix, IOContext context) {
+            throw new UnsupportedOperationException("this directory is read-only");
+        }
+
+        @Override
+        public void sync(Collection<String> names) {
+            throw new UnsupportedOperationException("this directory is read-only");
+        }
+
+        @Override
+        public void syncMetaData() {
+            throw new UnsupportedOperationException("this directory is read-only");
+        }
+
+        @Override
+        public void rename(String source, String dest) {
+            throw new UnsupportedOperationException("this directory is read-only");
+        }
+
+        @Override
+        public void close() {
+            // no-op
```

Comment:
I wonder if this should set `isOpen=false`? But I also wonder if we should not just extend `Directory` rather than `BaseDirectory`, the locking in `BaseDirectory` seems unnecessary here? None of this is really important though.